File: Medium/1282. Group the People Given the Group Size They Belong To/1282. Group the People Given the Group Size They Belong To.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<List<Integer>> groupThePeople(int[] groupSizes) {
        Map<Integer, List<Integer>> groups = new HashMap<>(); // Using a map to store groups
        List<List<Integer>> result = new ArrayList<>(); // Initialize the result list

        for (int i = 0; i < groupSizes.length; i++) {
            int size = groupSizes[i];

            if (!groups.containsKey(size)) {
                groups.put(size, new ArrayList<>()); // Initialize the group if it doesn't exist
            }

            groups.get(size).add(i); // Add the person to the corresponding group

            if (groups.get(size).size() == size) {
                // If the group is full, add it to the result and reset it
                result.add(groups.get(size));
                groups.put(size, new ArrayList<>());
            }
        }

        for (List<Integer> group : groups.values()) {
            if (!group.isEmpty()) {
                result.add(group); // Add any remaining groups to the result
            }
        }

        return result;
    }
}

File: Medium/1282. Group the People Given the Group Size They Belong To/1282. Group the People Given the Group Size They Belong To.rs
impl Solution {
    pub fn group_the_people(group_sizes: Vec<i32>) -> Vec<Vec<i32>> {
        let mut groups = vec![Vec::new(); 501]; // Initialize an array of empty vectors

        for (i, size) in group_sizes.iter().enumerate() {
            groups[*size as usize].push(i as i32); // Add the person to the corresponding group
        }

        let mut result = Vec::new();

        for (size, group) in groups.iter_mut().enumerate() {
            if !group.is_empty() {
                let mut start = 0;
                while start < group.len() {
                    let end = start + size;
                    result.push(group[start..end].to_vec());
                    start = end;
                }
            }
        }

        result
    }
}
